cuqi/experimental/mcmc/_laplace_approximation.py

```diff
@@ -0,0 +1,155 @@
+    ----------
+    target : `cuqi.distribution.Posterior`
+        The target posterior distribution to sample.
+
+    initial_point : ndarray
+        Initial parameters. *Optional*
+
+    maxit : int
+        Maximum number of inner iterations for solver when generating one sample.
+
+    tol : float
+        Tolerance for inner solver. Will stop before maxit if the inner solvers convergence check reaches tol.
+
+    beta : float
+        Smoothing parameter for the Gaussian approximation of the Laplace distribution. Larger beta is easier to sample but is a worse approximation.
+
+    callback : callable, *Optional*
+        If set this function will be called after every sample.
+        The signature of the callback function is `callback(sample, sample_index)`,
+        where `sample` is the current sample and `sample_index` is the index of the sample.
+        An example is shown in demos/demo31_callback.py.
+    """
+    def __init__(self, target, initial_point=None, maxit=50, tol=1e-4, beta=1e-5, **kwargs):
+
+        # Parameters (beta is used in target setter)
+        self.maxit = maxit
+        self.tol = tol
+        self.beta = beta
+
+        super().__init__(target=target, initial_point=initial_point, **kwargs)
+
+        if initial_point is None: #TODO: Replace later with a getter
+            self.initial_point = np.zeros(self.dim)
+            self._samples = [self.initial_point]
+
+        self.current_point = self.initial_point
+        self._acc = [1] # TODO. Check if we need this
+
+    @property
+    def prior(self):
+        return self.target.prior
+
+    @property
+    def likelihood(self):
+        return self.target.likelihood
+
+    @property
+    def model(self):
+        return self.target.model     
+    
+    @property
+    def data(self):
+        return self.target.data
+
+    @SamplerNew.target.setter
+    def target(self, value):
+        """ Set the target density. Runs validation of the target. """
+        # Accept tuple of inputs and construct posterior
+        super(UGLANew, type(self)).target.fset(self, value)
+        self._precompute()
+
+    def _precompute(self):
+        # Extract diff_op from target prior
+        D = self.prior._diff_op
+        n = D.shape[0]
+
+        # Gaussian approximation of LMRF prior as function of x_k
+        def Lk_fun(x_k):
+            dd =  1/np.sqrt((D @ x_k)**2 + self.beta*np.ones(n))
+            W = sp.sparse.diags(dd)
+            return W.sqrt() @ D
+        self.Lk_fun = Lk_fun
+
+        self._m = len(self.data)
+        self._L1 = self.likelihood.distribution.sqrtprec
+
+        # If prior location is scalar, repeat it to match dimensions
+        if len(self.prior.location) == 1:
+            self._priorloc = np.repeat(self.prior.location, self.dim)
+        else:
+            self._priorloc = self.prior.location
+
+        # Initial Laplace approx
+        # self._L2 = Lk_fun(self.x0)
+        self._L2 = Lk_fun(np.zeros(self.dim)) #TODO: fix this
```

Comment:
I see the issue with the "_precompute" method here. I think we need to follow a design pattern similar to what Amal did in the NUTS sampler. Use a "_pre_sample" and "_pre_warmup" methods. Have these called before sampling to define parameters. See e.g. in https://github.com/CUQI-DTU/CUQIpy/pull/387

Reply:
Done as suggested

Reply:
Looks good to me. Leaving unresolved for next reviewer to see.